core/trino-main/src/main/java/io/trino/execution/SetTimeZoneTask.java

```diff
@@ -0,0 +1,210 @@
+import static io.trino.util.Failures.checkCondition;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class SetTimeZoneTask
+        implements DataDefinitionTask<SetTimeZone>
+{
+    private final SqlParser sqlParser;
+    private final GroupProvider groupProvider;
+    private final StatsCalculator statsCalculator;
+
+    @Inject
+    public SetTimeZoneTask(SqlParser sqlParser, GroupProvider groupProvider, StatsCalculator statsCalculator)
+    {
+        this.sqlParser = requireNonNull(sqlParser, "sqlParser is null");
+        this.groupProvider = requireNonNull(groupProvider, "groupProvider is null");
+        this.statsCalculator = requireNonNull(statsCalculator, "statsCalculator is null");
+    }
+
+    @Override
+    public String getName()
+    {
+        return "SET TIME ZONE";
+    }
+
+    @Override
+    public ListenableFuture<Void> execute(
+            SetTimeZone statement,
+            TransactionManager transactionManager,
+            Metadata metadata,
+            AccessControl accessControl,
+            QueryStateMachine stateMachine,
+            List<Expression> parameters,
+            WarningCollector warningCollector)
+    {
+        String timeZoneId = statement.getTimeZone()
+                .map(timeZone -> getTimeZoneId(timeZone, statement, metadata, accessControl, stateMachine, parameters, warningCollector))
+                .orElse(TimeZone.getDefault().getID());
+        stateMachine.addSetSessionProperties(TIME_ZONE_ID, timeZoneId);
+
+        return immediateVoidFuture();
+    }
+
+    private String getTimeZoneId(
+            Expression expression,
+            SetTimeZone statement,
+            Metadata metadata,
+            AccessControl accessControl,
+            QueryStateMachine stateMachine,
+            List<Expression> parameters,
+            WarningCollector warningCollector)
+    {
+        Session session = stateMachine.getSession();
+        Analysis analysis = analyzeStatement(statement, metadata, accessControl, stateMachine, parameters, session);
+        Scope scope = Scope.builder()
+                .withRelationType(RelationId.anonymous(), new RelationType())
+                .build();
+        ExpressionAnalysis expressionAnalysis = ExpressionAnalyzer.analyzeExpression(
+                session,
+                metadata,
+                groupProvider,
+                accessControl,
+                sqlParser,
+                scope,
+                analysis,
+                expression,
+                warningCollector,
+                CorrelationSupport.ALLOWED);
+        Expression rewrittenExpression = rewriteExpression(expression, analysis);
+        Type type = expressionAnalysis.getType(expression);
+        if (!(type instanceof VarcharType || type instanceof IntervalDayTimeType)) {
+            throw new TrinoException(TYPE_MISMATCH, format("Time Zone expression '%s' not supported", expression));
```

Comment:
The message should also be adjusted accordingly. It's not just that it's "not supported" (which conveys that it's not yet implemented). The type is incorrect. I would rephrase the message as something like: "Expression must be varchar or interval day-time type, but was xxxx: <expression>".

Reply:
You are right. Thanks for raising this point.

Reply:
For the statement

```
SET TIME ZONE e();
```

the user will see now:

```
Expected expression of varchar or interval day-time type, but 'e()' has double type
```